src/lib/foreign-field.ts

```diff
@@ -0,0 +1,496 @@
+      // TODO: can this be made more efficient? since the limbs are already range-checked
+      if (length === sizeInBits) x.assertValidElement();
+      return x;
+    }
+
+    // Provable<ForeignField>
+
+    /**
+     * `Provable<ForeignField>.toFields`, see {@link Provable.toFields}
+     */
+    static toFields(x: ForeignField) {
+      let [, ...limbs] = x.value;
+      return limbs.map((x) => new Field(x));
+    }
+
+    /**
+     * Instance version of `Provable<ForeignField>.toFields`, see {@link Provable.toFields}
+     */
+    toFields() {
+      return ForeignField.toFields(this);
+    }
+
+    /**
+     * `Provable<ForeignField>.toAuxiliary`, see {@link Provable.toAuxiliary}
+     */
+    static toAuxiliary(): [] {
+      return [];
+    }
+    /**
+     * `Provable<ForeignField>.sizeInFields`, see {@link Provable.sizeInFields}
+     */
+    static sizeInFields() {
+      return 3;
+    }
+
+    /**
+     * `Provable<ForeignField>.fromFields`, see {@link Provable.fromFields}
+     */
+    static fromFields(fields: Field[]) {
+      let fieldVars = fields.map((x) => x.value);
+      let limbs = arrayToTuple(fieldVars, 3, 'ForeignField.fromFields()');
+      return new ForeignField([0, ...limbs]);
+    }
+
+    /**
+     * `Provable<ForeignField>.check`, see {@link Provable.check}
+     *
+     * This will check that the field element is in the range [0, p),
+     * where p is the foreign field modulus.
+     *
+     * Exception: If {@link createForeignField} is called with `{ unsafe: true }`,
+     * we don't check that field elements are valid by default.
+     */
+    static check(x: ForeignField) {
+      // if the `unsafe` flag is set, we don't add any constraints when creating a new variable
+      // this means a user has to take care of proper constraining themselves
+      if (!unsafe) x.assertValidElement();
+    }
+  }
+
+  function toFp(x: bigint | string | number | ForeignField) {
+    if (x instanceof ForeignField) return x.toBigInt();
+    return mod(BigInt(x), p);
+  }
+  function toVar(x: bigint | number | string | ForeignField): ForeignFieldVar {
+    if (x instanceof ForeignField) return x.value;
+    return ForeignFieldVar.fromBigint(mod(BigInt(x), p));
+  }
+  function isConstant(x: bigint | number | string | ForeignField) {
+    if (x instanceof ForeignField) return x.isConstant();
+    return true;
+  }
+
+  return ForeignField;
+}
+
+// helpers
+
+const limbMax = (1n << limbBits) - 1n;
+
+// the max foreign field modulus is f_max = floor(sqrt(p * 2^t)), where t = 3*limbBits = 264 and p is the native modulus
+// see RFC: https://github.com/o1-labs/proof-systems/blob/1fdb1fd1d112f9d4ee095dbb31f008deeb8150b0/book/src/rfcs/foreign_field_mul.md
+// since p = 2^254 + eps for both Pasta fields with eps small, a fairly tight lower bound is
+// f_max >= sqrt(2^254 * 2^264) = 2^259
+const foreignFieldMaxBits = (BigInt(Fp.sizeInBits - 1) + 3n * limbBits) / 2n;
+const foreignFieldMax = 1n << foreignFieldMaxBits;
```

Comment:
Actually this is not super super accurate.

We computed with sage the real maximum possible field we can support with pallas and vesta which is the one with prime order `926336713898529563388567880069503262826888842373627227613104999999999999999607 `